spitwad/spitwad.6
.TH spitwad 6 "22 Feb 2012"
.SH NAME
spitwad \- list directory and get stats of WAD files
.SH SYNOPSIS
.B spitwad
.I wadfile
.SH DISCLAIMER
This program is furnished on an AS IS basis for the fun of it.
id Software, Raven Software and Activision do not support this program.
.SH DESCRIPTION
.B spitwad
prints the contents of the directory of a wad file (iwad or
pwad), plus some overall statistics about the wad file.

Before the lump listing: the filename, number of lumps in the file,
and offset of the WAD info table are printed.

Lump listing output is in table format. For each lump in the wad, the following
information is printed:
.TP
.B lump
lump index within the file. The first lump is index 0, the
rest are numbered sequentially. The last lump in a file with
.I
n
lumps is numbered
.I
n\-1
.TP
.B name
The lump name.
.TP
.B size
Lump size in bytes (decimal)
.TP
.B filepos
Lump offset from the start of the file, in bytes (decimal)
.TP
.B 0xsize
Same as \fBsize\fR, but displayed in hex.
.TP
.B 0xfpos
Same as \fBfilepos\fR, but displayed in hex.
.PP
There is 7th column with no header. This column will display \fB(lzss)\fR
for lumps compressed with LZSS compression (only supported in the Jaguar
console port of Doom). Otherwise, the column will be blank.

After the lump listing: file size, data size, info table size,
and wasted space are printed. File size should be equal to the sum
of the other 3 values.
.SH "SUPPORTED GAMES"
Doom, Ultimate Doom, Doom][, Final Doom, Heretic, Hexen, Strife.
.SH "SEE ALSO"
The Unofficial Doom Specs by Matthew S Fell.
.br
.BR xwadtools (6)
.BR lswad (6)
.br
The original version of spitwad at
.nh
http://www.gamers.org/pub/idgames/historic/dmutils.zip
.hy
.SH BUGS
.B spitwad
will accept two extra arguments (called \fIstart\fR and \fIend\fR
in the usage message), but it doesn't use them for anything.
.SH AUTHOR
.B spitwad
was written by John Carmack
.PP
Slightly modified by B. Watson (yalhcru@gmail.com), to get it to compile
and run on modern OSes, and to add PWAD file support. Also wrote this
man page.


ChangeLog
See "ChangeLog.old" for Udo's changes.

20120227 bkw:

- wadgc: delete temp files after closing them. They're still being
  created in the wrong place (current dir).

- dmpsmu:
  no longer prints high-ASCII garbage in the Dump command.
  now exits if EOF on stdin, like a proper UNIX program (among
  other things, this makes it more easily scriptable)

- testsuite, made a pretty good start on.

- Add - option to trigcalc (normal I/O rather than termios). Allow
  it to compile for mingw (acts like - always present).

- various minor man page fixes

20120223 bkw:

- Fixed FreeBSD 9.0 compile.

- Added crosswin target. Borrowed a fnmatch() implementation from
  OpenBSD, since mingw-w64 seems to lack one.

20120222 bkw:

- Rename Andre's strlcat() to str_lower_cat(). Some OSes (FreeBSD and
  the MinGW cross compiler I have) define a strlcat() that does
  something different.

  - Only include internal strcasecmp() and strncasecmp() if
  !defined(HAVE_STRCASECMP). I don't have an exhaustive knowledge of
  which platforms do and don't provide this function, so the Makefile
  will define HAVE_STRCASECMP by default, with a note to the user.

- lswad's -F now correctly accepts FF_START as start-of-flats marker. Also
  added a warning if -F is used, and there are flats, but no F_END.

- Upgraded deutex to Amajorel Andre's 4.4.902, from here:
  http://www.teaser.fr/~amajorel/deutex/fungus/deutex-4.4.902.tar.gz
  I've made only one minor change to his code: the default log file
  is now /dev/null (can still set it with -log option, if you really
  want).

- Added John Carmack's old spitwad tool, from dmutils.zip. Find the
  original version here:
  http://www.gamers.org/pub/idgames/historic/dmutils.zip

20120221 bkw:

- added a generic man page "doom.6". Too many source ports don't
  supply a man page. Some of them don't supply any documentation
  on command line options at all (or it'll only be documented on
  their web site).

- man pages now installed gzipped by default. "make install GZIP_MAN=no"
  disables.

- binaries are no longer linked with -s (stripped). However, they're
  installed that way by default. "make install STRIP_BINS=no" to
  install unstripped binaries.

20120220 bkw:

- wadcat, mkqmap, mkpopal, mkgipal, mkxppal, mktran were all
  segfaulting if their options that require an argument were given at
  the end of the command line (e.g. "mkqmap -c"), added NULL checks

20120219 bkw:

- New utilities wadwhich, wadconvgame, wadconvgfx

- "make install" now installs the docs (to DOCDIR)

- "make install" now installs mergewad and tkwadcad

20120215 bkw:

- All of Andre Amajorel's patches have been integrated.

- All the tools now compile and run correctly on x86_64 Linux (so far
  as I've been able to determine). This took some effort!

- Makefile.lin has been renamed to Makefile. The other Makefile.* remain,
  but they haven't been updated for this release. I will accept patches
  from anyone who wants to use this on non-Linux systems, but I don't
  have the resources to do any testing on anything but Linux.

- Makefile now supports PREFIX and DESTDIR, should make life easier for
  packagers.

- XForms is now optional, can build with "make USE_XFORMS=no". It's only
  used for xew, so xew won't be built if XForms is disabled.

- The acc utility is now installed as acc-raven, to avoid conflicts with
  the acc utilities provided by VaVoom and ZDoom.

- libfindwad: new library that searches for WAD files in a uniform manner.
  Tools are slowly being updated to use it. See ENVIRONMENT section of
  xwadtools manpage for usage details; see include/findwad.h for API
  and lib/findwad.c for implementation.
  Currently these tools support libfindwad:
  dm2au  wadps  wadwhat  lswad  wadsprit  wadpat  wadtex  wadext
  wadtxls  wadflat

- dm2au:
  New options -r (raw audio output) and -w (wav output).
  Output file can be specified, and can be - for standard output.
  Man page contains a few examples for using dm2au with sox.

- New utility midi3mus. Converts MIDI files to MUS lumps. Written by natt
  on the DoomWorld forums, thread:
  http://www.doomworld.com/vb/doom-editing/55814-midi2mus-64-bit-compatible/

- slige now falls back to $LIBDIR/slige/slige.cfg if no -config or
  $SLIGECONFIG is given and if slige.cfg doesn't exist in the current
  directory.

- wadps: Default game changed to doom2. Behaviour for
  doom II wads now matches doom I, when no level is selected (all maps
  are included in the output; previously the output was just empty).

- lswad: Added --help/-? option. Needed because its default action is now
  to list the contents of the default IWAD.

- wadtxls: Output now sent to stdout if no -o is given.
  Episode and Mission (-e and -m) arguments are no longer required; they
  default to E1M1 for doom 1 and MAP01 for doom 2.

- raw2sfx: outfile argument can be - for stdout. Warn if there's a RIFF or id
  sfx header, or if the input's really short.

- various compiler warnings have been squelched. Quite a few are left.



trunk/spitwad/spitwad.c
#define VERSION	"1.1"

/*
=============================================================================

							  SPITWAD

						  by John Carmack

=============================================================================
*/


#include <stdint.h>

#include "cmdlib.h"

#include "sysdep.h"
#define LittleLong(x) bswapl(x)

//================
//
// wad file types
//
//================

typedef struct
{
	uint32_t		filepos;
	uint32_t		size;
	char		name[8];
} lumpinfo_t;


typedef struct
{
	char	identification[4];
	uint32_t	numlumps;
	uint32_t	infotableofs;
} wadinfo_t;


/*
===============
=
= main
=
===============
*/

int main (int argc, char **argv)
{
	char	filename[128];
	int		handle;
	wadinfo_t	wad;
	lumpinfo_t	*lumps, *lump_p;
	int		i,j;
	char	name[9];
	uint32_t	datasize, wadsize, infosize;
	char	*lzss;


	printf ("SPITWAD "VERSION" by John Carmack, copyright (c) 1992 Id Software\n");

	if (CheckParm ("?") || (argc != 2 && argc != 4))
	{
		printf ("SPITWAD filename<.wad> [start end]\n");
		exit (1);
	}

	strcpy (filename, argv[1]);
	DefaultExtension (filename, ".wad");

	handle = SafeOpenRead (filename);
	wadsize = filelength(handle);
	datasize = 0;

	SafeRead (handle, &wad, sizeof(wad));

	if (strncmp(wad.identification,"IWAD",4) && strncmp(wad.identification,"PWAD",4))
		Error ("Wad file %s doesn't have IWAD or PWAD id\n",filename);

	wad.numlumps = LittleLong(wad.numlumps);
	wad.infotableofs = LittleLong(wad.infotableofs);

	infosize = wad.numlumps*sizeof(lumpinfo_t);
	lumps = lump_p = SafeMalloc(infosize);
	lseek (handle, wad.infotableofs, SEEK_SET);
	SafeRead (handle, lumps, infosize);
	infosize += sizeof(wadinfo_t);

	printf ("\n");
	printf ("WAD file    : %s\n",filename);
	printf ("numlumps    : %u\n",wad.numlumps);
	printf ("infotableofs: %u\n",wad.infotableofs);
	printf ("\n");
	printf ("lump name       size  filepos    0xsize 0xfpos\n");
	printf ("---- -------- ------ --------    ------ ------\n");

	name[8] = 0;

	for (i=0 ; i<wad.numlumps ; i++, lump_p++)
	{
		datasize += LittleLong (lump_p->size);
		strncpy (name,lump_p->name,8);
		for (j=7 ; j>=0 ; j--)
			if (!name[j])
				name[j] = ' ';
			else
				break;
		if (name[0] & 0x80)
		{
			lzss = "(lzss)";
			name[0] &= 0x7f;
		}
		else lzss = "";
		printf ("%4i %s %6i %8i    %6x %6x %s\n", i, name
		, LittleLong(lump_p->size), LittleLong(lump_p->filepos)
		, LittleLong(lump_p->size), LittleLong(lump_p->filepos),
		lzss );
	}

	printf ("\n");
	printf ("File size:%9i\n", wadsize);
	printf ("Data size:%9i\n", datasize);
	printf ("Info size:%9i\n", infosize);
	printf ("Wasted   :%9i\n", wadsize-(datasize+infosize));

	return 0;
}
